algorithms/random.go

```diff
@@ -0,0 +1,72 @@
+package algorithms
+
+import (
+	"math/rand"
+	"sync"
+	"time"
+)
+
+var rng = struct {
+	sync.Mutex
+	rand *rand.Rand
+}{
+	rand: rand.New(rand.NewSource(time.Now().UnixNano())),
+}
+
+// Int returns a non-negative pseudo-random int.
+func Int() int {
+	rng.Lock()
```

Comment:
why do we need to mutex these funcs?

Reply:
To ensure that a value is not accessed concurrently. Generally, it's a thread-safe version.

Reply:
Oh apparently using rand.NewSource makes it un-threadsafe, weird.